### ISSUE TITLE
 Scan WhatsApp web barcodes remotely

### ISSUE THREAD

**Issue by Nottt** [BenSchZA/docker-android-avd#4]
This would be very great. 

Do you have any ideas on how to implement this? I assume you'll use ffmpeg? I could try to look at this

**Comment by BenSchZA** [comment]
Hey @Nottt ! I'd be keen to pick this up again and see if you can help. It was a combination of ffmpeg and v4l2loopback (https://github.com/umlaeute/v4l2loopback) as far as I can remember. I'll try find the WIP script I had set up.

```I'd like to stream a VNC feed via the v4l2loopback kernel module into the Android AVD... WIP. This would allow one to scan WhatsApp web barcodes remotely.```

**Comment by blackchineykh** [comment]
> Hey @Nottt ! I'd be keen to pick this up again and see if you can help. It was a combination of ffmpeg and v4l2loopback (https://github.com/umlaeute/v4l2loopback) as far as I can remember. I'll try find the WIP script I had set up.
> 
> `I'd like to stream a VNC feed via the v4l2loopback kernel module into the Android AVD... WIP. This would allow one to scan WhatsApp web barcodes remotely.`

Hi, was this ever working? I was keen to look at this as well

**Comment by SophiaHadash** [comment]
Any luck so far? I really need a Whatsapp server, but currently I am stuck at this part:

![image](https://user-images.githubusercontent.com/5069005/102384306-dc7d2780-3fcc-11eb-8c93-b437d37b8034.png)

Would be awesome if we can feed an image containing the QR code to the device camera feed.

**Comment by blackchineykh** [comment]
> Any luck so far? I really need a Whatsapp server, but currently I am stuck at this part:
> 
> ![image](https://user-images.githubusercontent.com/5069005/102384306-dc7d2780-3fcc-11eb-8c93-b437d37b8034.png)
> 
> Would be awesome if we can feed an image containing the QR code to the device camera feed.

What are you using for the emulator? I was thinking if a remote stream could be sent to the back camera then this would be the solution

**Comment by SophiaHadash** [comment]
I have a remote x86 server that runs docker. Launched an android image using docker-compose and connect to it using vnc. However, vnc input does not work so I used rulers (KRuler) and adb commands to provide input.

```
adb connect remote
adb shell input keyevent <key> # for typing
adb shell input tap <x> <y> # for key presses
```

**Comment by SophiaHadash** [comment]
I will try tonight to use `v4l2loopback` to feed Whatsapp the QR code. Gonna follow this guide:
https://stackoverflow.com/questions/6165618/how-to-mock-a-picture-in-android-emulator-camera#35526103

I will report whether it works, maybe make a pull request.

**Comment by xeniter** [comment]
used this but for android avd emulator:
https://github.com/tulir/mautrix-whatsapp/wiki/Android-VM-Setup
my plan is still to move whatsapp into a jail on a server, will try this docker image

**Comment by cosmo83** [comment]
I coukd get it running, not with this container, but my guess is that this shoudl work

![image](https://user-images.githubusercontent.com/973160/102710816-173bd400-42db-11eb-9169-83c737095636.png)

**Comment by smashah** [comment]
Thanks for the recent discussions everyone. It's no secret this repo is a bit stale and needs a lot of community effort. Very open to accepting PRs. once the holiday season is over I'll be spending more time on this project. Of there is a different method to get this working then we should go with that as the goal of this project is to get an avid working in a docker container

**Comment by xeniter** [comment]
hi, created PR, don't know if you would like to have everything from it:
https://github.com/open-wa/wa-avd-docker/pull/13
at least it was possible to scan qr code from browser in container, however you need the v4l2loopback loaded already on host (modprobe v4l2loopback)

**Comment by 0xhex** [comment]
@cosmo83 hello, can you please reach me?
i have paid job about it, can you install same thing on my computer?
efekanpulatli@gmail.com

**Comment by xeniter** [comment]
Yes tell your friends using signal and get rid of the data stealing messenger.

if they are still ignorant here is my PR:
https://github.com/open-wa/wa-avd-docker/pulls 
you have to "modprobe v4l2loopback" on the host to get video running
after that start android with camera support /app/start-avd-with-cam.sh and scan the QR from desktop opened in the browser for example

**Comment by laruche** [comment]
I've already use your pr :) and its work like a charm
But i would like to use it on cloud :) and my ovh server doesn't have usb port :D 
So @cosmo83 solution look like intersting :D 

About Signal.... 
In Morocco is really complicated because whatsapp is our SMS :/ 
And people use it since Whatsapp beginning, they don't care about there data 
I work with Local Artisan, mostly illiterate
So install new app, no way :D

**Comment by mzalazar** [comment]
I made it!!!! using l4l2loopback =)

![Screenshot at 2021-07-16 11-52-13](https://user-images.githubusercontent.com/942079/125967295-faace5ab-8cc8-4b40-8553-b1d7d1875874.png)

**Comment by blackchineykh** [comment]
> I made it!!!! using l4l2loopback =)
> 
> ![Screenshot at 2021-07-16 11-52-13](https://user-images.githubusercontent.com/942079/125967295-faace5ab-8cc8-4b40-8553-b1d7d1875874.png)

Thats awesome news! Can you send how were you able to achieve this? I need to do the same

**Comment by mzalazar** [comment]
I used some "crazy stuff" (using ffmpeg) like cosmo83 said... the TRICKY part is to generate the qr image, I'm using @open-wa/wa-automate and everytime I launch the program (logged out of course) it generates a QR code, I save that code into a image png and I send that image to my v4l2loopback device. Please excuse my english... only basic =)
